### PR TITLE
Rewording and clarifying added missing properties

### PR DESCRIPTION
**Pull request by matthiashanel**
Also described message flow in interest propagation section

Signed-off-by: Matthias Hanel <mh@synadia.com>

### CODE REVIEW

File: nats-server/configuration/gateways/README.md [hunk 1]
```diff
@@ -2,26 +2,30 @@
...
-* gateways gossip gateway members and remote discovered gateways
+* gateways gossip gateway nodes and remote discovered gateways
 
 Unlike clusters, gateways:
 
-* don't form a full mesh
+* have names, specifying the cluster they are in
+* don't form a full mesh between gateway nodes, but form a full mesh between cluster instead
 * are bound by uni-directional connections
+* don't gossip gateway nodes to clients
```
Comment (kozlovic) [inline nats-server/configuration/gateways/README.md]: I think I know what you mean here, but to be clear, nodes from a cluster are not gossiped to clients connected to another cluster. That is, if cluster A has 3 servers A1, A2 and A3, and cluster B has 2 servers B1 and B2, a client connecting to A1 will be notified of A2 and A3, but no server from B. Same for cluster B, a client connecting to B1 will be notified of B2 but no server from cluster A.

File: nats-server/configuration/gateways/README.md [hunk 2]
```diff
@@ -2,26 +2,30 @@
...
 
 Gateways exist to:
 
 * reduce the number of connections required between servers 
 * optimize the interest graph propagation
 
+If gateways are to be used in a cluster, **all** server of this cluster need to have a gateway configuration with the **same name**. Furthermore, every gateway node needs to be able to **connect to every** other gateway node and vice versa. Everything else is considered a miss configuration.
```
Comment (kozlovic) [inline nats-server/configuration/gateways/README.md]: > be able to **connect to every** other gateway node

That could be interpreted as a node in cluster A needs to be able to connect to all nodes in cluster B, as in creating a connection to all nodes. As you know, this is not the case. Each node needs to be able to "connect to any" other gateway node.

> miss configuration

I think "misconfiguration" instead?

File: nats-server/configuration/gateways/README.md [hunk 3]
```diff
@@ -2,26 +2,30 @@
...
+
 ## Gateway Connections
 
-A nats-server in a gateway role will specify a port where it will accept gateways connections. If the configuration specifies other _external_ `gateways`, the gateway will create one outbound gateway connection for each gateway in its configuration. It will also gossip other gateways it knows or discovered.
+A nats-server in a gateway role will specify a port where it will accept gateway connections. If the configuration specifies other _external_ `gateways`, the gateway will create one outbound gateway connection for each gateway in its configuration. It will also gossip other gateways it knows or discovered. Fewer _external_ `gateways` mean less configuration. Ye, the ability to discover more gateways and gateway nodes depends on these server running. This is similar to _seed server_ in cluster. It is recommended to have all _seed server_ of a cluster listed in the `gateways` section.
```
Comment (kozlovic) [inline nats-server/configuration/gateways/README.md]: > other gateways it knows or discovered. 

or "discovers" to be in the same tense?

> Ye, the ability to discover

Yet?

File: nats-server/configuration/gateways/README.md [hunk 4]
```diff
@@ -70,7 +77,7 @@ When a gateway on _A_ sends many messages on various subjects for which _B_ has
 
 When a queue subscriber creates a new subscription, the gateway propagates the subscription interest to other gateways. The subscription interest is only propagated _once_ per _Account_ and subject. When the last queue subscriber is gone, the cluster interest is removed.
 
-Queue subscriptions work on _Interest-only Mode_ to honor NATS' queue semantics across the _Super Cluster_. For each queue group, a message is only delivered to a single queue subscriber. Only when a local queue group member is not found, is a message forwarded to a different interested cluster; gateways will always try to serve local queue subscribers first and only failover when a local queue subscriber is not found.
+Queue subscriptions work on _Interest-only Mode_ to honor NATS' queue semantics across the _Super Cluster_. For each queue group, a message is only delivered to a single queue subscriber. Only when a local queue group member is not found, is a message forwarded to a different interested cluster; gateways will always try to serve local queue subscribers first and only failover when a local queue subscriber is not found. The interested cluster is picked based on lowest RTT.
```
Comment (kozlovic) [inline nats-server/configuration/gateways/README.md]: > Only when a local queue group member is not found, is a message forwarded to a different interested cluster; gateways will ...

Would it read better like this?
"If the same queue group exists in multiple clusters, the server will pick one member from the queue group in its cluster, only sending to a different cluster if there is no interest in its cluster. In other words, a server will always try.."

File: nats-server/configuration/gateways/README.md [hunk 5]
```diff
@@ -48,14 +52,17 @@ The number of connections required to join clusters using clustering vs. gateway
...
 
-Gateways propagate interest using three different mechanisms:
+## Interest Propagation
+Messages from clients directly connected to a gateway node will be sent along outgoing gateway connections according to the following three interest propagation mechanisms:
 
 * Optimistic Mode
 * Interest-only Mode
 * Queue Subscriptions
 
+Local interest permitting, the receiving gateway node sends the messages directly to it's subscribing clients as well as server within the cluster that have subscribing clients.
```
Comment (gcolliso) [inline nats-server/configuration/gateways/README.md]: ...to it's subscribing clients as well as <any><all><a> server within the cluster<s> that have ...

File: nats-server/configuration/gateways/README.md (hunk 1 above)
Comment (matthiashanel) [reply nats-server/configuration/gateways/README.md]: This in the section where gateways are contrasted with cluster
Unlike clusters, gateways:
...
* don't gossip gateway nodes to clients

That is essentially what you are saying.

File: nats-server/configuration/gateways/README.md (hunk 2 above)
Comment (matthiashanel) [reply nats-server/configuration/gateways/README.md]: changed to any and misconfiguration

File: nats-server/configuration/gateways/README.md (hunk 3 above)
Comment (matthiashanel) [reply nats-server/configuration/gateways/README.md]: fixed

File: nats-server/configuration/gateways/README.md (hunk 4 above)
Comment (matthiashanel) [reply nats-server/configuration/gateways/README.md]: agreed, also changed the last sentence to:
. The server will pick the cluster with the lowest RTT.

File: nats-server/configuration/gateways/README.md (hunk 5 above)
Comment (matthiashanel) [reply nats-server/configuration/gateways/README.md]: removed

File: nats-server/configuration/gateways/README.md [hunk 6]
```diff
@@ -70,7 +77,7 @@ When a gateway on _A_ sends many messages on various subjects for which _B_ has
...
 
-Queue subscriptions work on _Interest-only Mode_ to honor NATS' queue semantics across the _Super Cluster_. For each queue group, a message is only delivered to a single queue subscriber. Only when a local queue group member is not found, is a message forwarded to a different interested cluster; gateways will always try to serve local queue subscribers first and only failover when a local queue subscriber is not found.
+Queue subscriptions work on _Interest-only Mode_ to honor NATS' queue semantics across the _Super Cluster_. For each queue group, a message is only delivered to a single queue subscriber. If the same queue group exists in multiple clusters, the server will pick one member from the queue group in its cluster, only sending to a different cluster if there is no interest in its cluster. In other words, a server will always try to serve local queue subscribers first and only failover when a local queue subscriber is not found. The server will pick the cluster with the lowest RTT.
```
Comment (kozlovic) [inline nats-server/configuration/gateways/README.md]: >  only sending to a different cluster if there is no interest in its cluster.

After re-reading it, `in its cluster` seem redundant. So simply `if there is no local interest`. It's just that I hope user do not interpret "local" with server-only. In the context of gateways and subscription interest, here "local" means all nodes in a given cluster.

File: nats-server/configuration/gateways/README.md [hunk 7]
```diff
@@ -2,26 +2,30 @@
...
+
 ## Gateway Connections
 
-A nats-server in a gateway role will specify a port where it will accept gateways connections. If the configuration specifies other _external_ `gateways`, the gateway will create one outbound gateway connection for each gateway in its configuration. It will also gossip other gateways it knows or discovered.
+A nats-server in a gateway role will specify a port where it will accept gateway connections. If the configuration specifies other _external_ `gateways`, the gateway will create one outbound gateway connection for each gateway in it's configuration. It will also gossip other gateways it knows or discovers. Fewer _external_ `gateways` mean less configuration. Yet, the ability to discover more gateways and gateway nodes depends on these servers running. This is similar to _seed server_ in cluster. It is recommended to have all _seed server_ of a cluster listed in the `gateways` section.
```
Comment (kozlovic) [inline nats-server/configuration/gateways/README.md]: > the gateway will create one outbound gateway connection for each gateway in it's configuration.

@gcolliso I think the original `its` was correct. The meaning here is "possessive", so "its" is the correct word.

**Pull request merged**